### ISSUE TITLE
Allow support of TransA='T' in GEMMs

### ISSUE THREAD

**Issue by cgk-x** [hfp/libxsmm#366]
It appears to me that at this moment (clone of github master version from a few days ago), libxsmm will simply bypass all requests for GEMM kernels which have the LIBXSMM_GEMM_FLAG_TRANS_A flag set. For example, any
```
typedef libxsmm_mmfunction<double> kernel_type;
kernel_type kernel(LIBXSMM_GEMM_FLAG_TRANS_A, N,M,L, 1.0/*alpha*/, 0.0/*beta*/);
```
will return a 0-pointer kernel object. My reading of the macro LIBXSMM_GEMM_NO_BYPASS in libxsmm_generator.h is that this behavior is intentional, so I guess the corresponding kernel routines are simply not there and therefore cannot be provided to the calling program.

Maybe I am terribly confused, but I was of the impression that this was the easiest case to get right with high efficiency... which is precisely why in the quantum chemistry program I am working with most data is computed/aligned in such a way that it can be multiplied as TransA=T and TransB=N. And this, unfortunately, makes the current libxsmm inapplicable, although I think it could really help. 

I was wondering: Did I miss something and there already is a way to use the TN kernels? If not, are there any plans to provide the TN kernels in the foreseeable future?

**Comment by hfp** [comment]
Indeed, see https://github.com/hfp/libxsmm/wiki/Q&A#what-is-a-small-matrix-multiplication

> NOTE: Alpha, Beta, and TransA are limited to `1`, `{ 1, 0 }`, and `'N'` respectively.

**Comment by cgk-x** [comment]
As mentioned, I did not only read the documentation (which, as you say does state this limitation), but also the code. After first missing this side note.

What I am wondering is this:
- The TransA=T case is _very_ important in practical applications. Particularly in the case of small MxMs. As mentioned, I have seen many programs go out of their way to formulate intermediates  specifically in such a way that they _can_ be multiplied as TransA=T and TransB=N, because most BLAS implementations are most efficient in this case. Is there a specific reason libxsmm chose to simply not allow this case? Or is this just a "we had more important things to do"-matter?
- In the latter case, are there any plans to address this and add the TN kernels?

**Comment by hfp** [comment]
Do you have an application that you can share/describe? A reproducer may be helpful if it is a known/common application that has some else around plain matrix multiply. If it's plain matrix multiply, a list of parameters can help us to focus our effort (value ranges or specific cases for M, N, K, LDx, TransA/B, etc.).

**Comment by alheinecke** [comment]
Let me add some clarification. The statement which was made earlier is not true: TN is simple and very fast, it's the opposite.

TransA=T and TransB=N is the case which is horrible as it cannot be vectorized efficiently and that is true for any BLAS implementation as it's a hardware limitation. There is simply no hardware support for doing this (we would need 0.5 cycle throughput gather, today this is at 8 cycles) and therefore it's slow for all implementations. Of course this statement is for col-major data.

Here is the table:
NN: fast in rowmajor and colmajor
NT: fast in colmajor, super slow in rowmajor
TN: super slow in colmajor, fast in rowmajor
TT: OK speed in rowmajor and colmajor: we use this trick C=A^T * B^T ~ C^T = B * A

Can you therefore please clarify if you use row- or colmajor data? If you use rowmajor, the B^T is the option you are looking for the A^T matmul.

If you need A^T in col-major: please use our fast JITed transposes first to create a A_tmp and then call libxsmm GEMM with A not being transposed. This is the fastest implementation possible. For increasing speed, you should block looks in your application in such a way that you reuse this A_tmp as much as possible to amortize the transpose cost. 

We might add in future an A^T kernel, but this has to do exactly the same as mentioned in the previous paragraph (inlined transpose, using gather, etc.), so it's not faster, just more convenient but forbids reuse of A_tmp as we would transpose over and over again the same data. If A_tmp is reused multiple times, the over speed is close to ideal.

**Comment by cgk-x** [comment]
Dear alheinecke,
I am sorry, but are you sure we are talking about the same thing? The operation I mean with TN is what a traditional netlib blas dgemm would call TransA='T' and TransB='N' (with Fortran-like col-major nomenclature for BLAS stuff). In this case dimension being contracted over is linear with unit stride for both A and B. Written in pseudo-code for array data access, the TN operation would be this:
```
C[m + ldc*n] = sum_k A[k + m*lda] * B[k + n*ldb]
```
Now I did not check how you construct your MxM kernels, and it has been more than a decade that I last looked into the details of such things. But if A and B and their strides lda and ldc are aligned to suitable boundaries, it appears to me that one could directly do aligned loads of k-packets of data here (e.g., a 256bit 4-double AVX2 packet for k=0,k=1,k=2,k=3 in one go). And it would appear to me that this is a good thing. 

I think we might be either stumbling over some clash in nomenclature (but I even looked up the netlib reference on dgemm again (http://www.netlib.org/lapack/explore-html/d7/d2b/dgemm_8f.html) in case I ran into some major sanity slippage, since I rarely talk to DGEMM directly, and I see no way in which TN could mean anything else than what I just wrote)... or MxM kernels must work much different now from I imagine they do.

In this case, my updated questions:
- Is the operation I wrote the one you meant with TN in col major?

- In case we have a nomenclature clash: Can the operation I described there (i.e., entries C are given by dot products of column vectors of A and B, which lie in continuous memory, respectively) be done with libxmm? What would I have to ask it to get it? (since it's apparently not TN)

- Irrespectively of whether some operations are slower or less slow: The reason the traditional DGEMM provided all eight variants of transposing matrices (TransA and TransB directly, and TransC implicitly over (AB)^T = B^T A^T) was that all those cases did tend to occur in large scientific applications (of course, some like TN and NN more often than others). For example, commonly the matrices A would represent some sort of vector space, and one would need to transform data into and out of it (so need both A and A^T). So having the operation supported, even if it is slow (e.g., doing the explicit transpose you mention) may still be very relevant to make the software more applicable. Do you have plans to add such compatibility layers? Another one would be the alpha != 1 case, which I understand is also not really all that helpful from a kernel viewpoint, but quite frequently used, too.  The advantage of adding such things in the interface, with a compatibility layer if necessary, is that *if* there ever comes a point in which they can be done efficiently, the library has the required information (this is one of the reasons why in quantum chemistry we nowadays still have some programs which are running at 80-90% peak flops, despite being written in bad Fortran in the 1980s.).

**Comment by alheinecke** [comment]
Yes, MxM must work differently to get good performance. For performance it needs to be formulated as outer products and not inner products. 

What you wrote is exactly TN in colmajor, but there are the two problems:
- as the inner product needs full vectors for both A and B, we can no longer block efficiently for L1 cache (we now need 2 SIMD full loads per cycle, but L2 can only deliver 0.8 CLs per cycle on Skylake) -> we are below 50% peak when we fall out of L1, of course we can block differently, to mitigate the problem but it's not optimal
- we need horizontal SIMD operations. The inner product has a scalar as result, but when vectorizing an inner produce in the way you have described we need a fix-up sequence. This needs log_2(SIMD-width) ADDs and at least as much shuffles and then we need to shuffle the reduced values back into partial SIMD registers to perform masked stores. With AVX512 and FP64 we have 8 elements per vector, so if K=32-48 (this is what is small and where LIBXSMM shines), we have 4-6 FMADDs and the inner product is done. To hide FMA latencies we need to work at 16-24 innner product in parallel. That leaves us with a 6x4 result matrix. However, we hold that result matrix in a tensor which is 6x4x8 and we now need to reduce in z-dimension. We tried this 2 years ago and the overhead is in nearly all of the small GEMM cases much bigger than the actual inner product computation. Therefore the TN case is the worst. 

For NN and NT we can do outer products: We load of full vector of A (this is multiple rows of a given column) and multiply this with a SIMD-broadcasted scalar from B. That means we are now computing #SIMD-length inner products at the same time and we can do a simple full vector stores to C. Broadcast is much faster than reduction as it's parallel and not sequential by nature. Additionally, as the access to B is scalar, NN and NT achieve the same performance. Last but not least, when going back to our 6x4 blocking: we do 4 full vector loads and 6 broadcasts. So instead of 24 CL, the outer product kernel needs less than 5 CLs, so 5x less bandwidth pressure. 

We "support" TN as mentioned before: we have extremely fast transposes as JIT kernels: https://github.com/hfp/libxsmm/blob/master/src/template/libxsmm.h#L406 and https://github.com/hfp/libxsmm/blob/master/include/libxsmm_typedefs.h#L523 as we specialize into small GEMMs and tensor contractions, the idea is
- to create an A^T on the stack in user code 
- call the JITed transpose to fill A^T from A
- call the NN or NT kernel
- try to reuse A^T for other multiplications (with every reuse overall application performance will improve)

We plan to add this "crappy" flow in the kernel, however since we have the SW-based solution mentioned here, it's super-low priority and just a must-have for 2.0 which has an anticipated release Mid/End of this year. Also when fusing the transpose, the aforementioned amortization ability is lost. 
@hfp: we might want to add this flow to the documentation for TN...

Similar things are true for alpha != 1. It can be done with upfront scaling and normally there is reuse in tensor contractions, so it's fine. Again it's a must-have for 2.0, but we don't have a release date for this release:-)

Of course there is always the possibility to use libxsmm_dgemm which falls back to your BLAS implementation of choice if we cannot JIT and we could add the transpose trick here (if it's not already done this way, @hfp is maintaining this interface). Again the amortization possibility will be lost.

Last but not least, we are happy to work with you and we always do close collaborations. Therefore could you please share: precision, M, N, K, lda, ldb, ldc, col/row major, alpha/beta and transA/B of all the case you are interested in and how much runtime they take today in the full application. What's the full application? Then we can come up with a prioritization list.

**Comment by cgk-x** [comment]
Dear alheinecke,
thanks a lot for the explanation! I understand what is going on now, and what the reasoning for these choice are, and that they are necessary. I had failed to notice the change from inner-product based MxM kernels to outer product based kernels due to the FMADD business and horizontal SIMD, which in retrospect I really should have considered.  I am also very happy that I asked, because I now know that we want to go TransA=N (as mentioned, in older architectures from 32bit x86 to pre-AVX times to strange Unixes this was still different, and many programs went out of their way to formulate tasks as NT; and I'd have continued doing that unless I had asked. So thanks! This was a great help, and I appreciate you taking the time.).

Regarding the matter at hand:

- I agree that the solution you have, with the JIT transpose , is workable. I will play around with it a bit.

- Regarding the program I am working on: At this moment I just tried libxsmm on a proof-of-concept program to illustrate some quantum chemical molecular integration technique. This application is not exciting or important enough to deal with in terms of the library (it will in its current form likely never be used in actual production code), but I used it to get some ideas regarding whether I like libxsmm or not. I like to try new techniques.

- However, I am quite happy with the performance I see, and I am considering a test integration into the Molpro quantum chemistry package (www.molpro.net), where I'd use it as computational kernel in some sort of ``local correlation methods'', which do involve very large numbers of MxMs of vastly different sizes---including many small ones. This _would_ be important enough to care (molpro is installed on a large fraction of HPC clusters in the world and in those used to do high-accuracy thermochemical calculations by several thousand researchers). I will play around with it once I get to it and try to create a meaningful minimal test case (similar to the CP2K one you got) to get a better impression of how it works and what statistics on the core cases are.

Thanks again!

**Comment by hfp** [comment]
@cgk-x Let me know if/when you need some assistence wrt Molpro integration or whatever.

For the time being, I will close this issue and let you do your work or reproducer (if it plays out). Also, feel free to open an issue for any hurdle you may run into!
